### PR TITLE
GSAGH-530: Support 3 or more vertices for a Load Panel

### PR DESCRIPTION
**Pull request by SandeepArup**

### CODE REVIEW

File: GsaGH/Components/1_Properties/Create2dProperty.cs [hunk 1]
```diff
@@ -46,6 +46,8 @@ private readonly IReadOnlyDictionary<SupportType, string> _supportDropDown
         { SupportType.TwoAdjacentEdges, "Two adjacent edges" },
         { SupportType.OneEdge, "One edge" },
         { SupportType.Cantilever, "Cantilever" },
+         { SupportType.OneWay, "One-way" },
+        { SupportType.TwoWay, "Two-way" },
```
Comment (psarras) [inline GsaGH/Components/1_Properties/Create2dProperty.cs:50]: Remember to add a warning when any other type is used

File: GsaGH/Components/2_Geometry/Create2dElement.cs [hunk 2]
```diff
@@ -46,7 +50,30 @@ protected override void SolveInstance(IGH_DataAccess da) {
...
+          }
+          else {
+            throw new ArgumentException("Unable to retrieve polylines from the given curve");
+          }
+          break;
+        default: {
+            throw new ArgumentException("Input geometry is not supported to create a 2D element.");
+          }
+      }
+
```
Comment (psarras) [inline GsaGH/Components/2_Geometry/Create2dElement.cs:76]: extract all of that to a GeometryValidation method that outputs a GsaElement2d.

The Exceptions, need to be set as a new class extending the closest Exception Type.

For instance

```csharp
public class InvalidGeometryForProperty : Exception
{
    public InvalidGeometryForProperty(GH_Curve curve): Exception("..."){}
    // You can under other constructors here for the other exceptions
   // Or create a new class, if they don't really fit
  
}
```

File: GsaGH/Components/2_Geometry/Create2dElement.cs [hunk 3]
```diff
@@ -46,7 +50,30 @@ protected override void SolveInstance(IGH_DataAccess da) {
...
+      switch (geometry_parameter) {
+        case GH_Mesh mesh:
+          elem = new GsaElement2d(mesh.Value, isLoadPanel);
+          break;
+        case GH_Curve curve:
+          if (!isLoadPanel) {
+            throw new ArgumentException("Specify mesh geometry as the input parameter to create a 2D element.");
+          }
+          if (curve.Value.TryGetPolyline(out Rhino.Geometry.Polyline polyline)) {
+            if (polyline.ToArray().Length < 3) {
```
Comment (psarras) [inline GsaGH/Components/2_Geometry/Create2dElement.cs:63]: Add a GH Warning

DA.SetWarning...

To mention that an attempt was made to automatically convert Curve to Polyline.

That being said, can you check what happens if the user provides a polyline? I think we need to make an attempt to first cast the Curve to the various other types: Polyline, PolyCurve and NurbsCurve, and handle each case separately. Check what currently happens if user provides a weird curve.

Comment (psarras) [reply GsaGH/Components/2_Geometry/Create2dElement.cs:63]: if it becomes too complex, perhaps we can offload the logic for extracting the polyline to the user and ask for a true polyline to be provided.

**Review by psarras — CHANGES_REQUESTED** [review]

File: GsaGH/Parameters/2_Geometry/GsaElementFactory.cs [hunk 4]
```diff
@@ -59,14 +95,13 @@ internal static ConcurrentBag<GsaElement2dGoo> CreateElement2dFromApi(
         ConcurrentDictionary<int, GSAElement> elems = sortedElements[parentId];
         var prop2Ds = new ConcurrentDictionary<int, GsaProperty2d>();
         var mList = new ConcurrentDictionary<int, Mesh>();
-
+        var polylineList = new ConcurrentDictionary<int, Rhino.Geometry.Polyline>();
```
Comment (DominikaLos) [inline GsaGH/Parameters/2_Geometry/GsaElementFactory.cs:98]: not used variable

File: GsaGHTests/3_Components/2_Geometry/CreateElement2dTests.cs [hunk 5]
```diff
@@ -13,27 +13,45 @@
...
+        new Point3d(1, 0, 0),
+        new Point3d(1, 1, 0),
+        new Point3d(0, 1, 0),
+      };
+        points.Add(points[0]);
+        var polyline = new Rhino.Geometry.Polyline(points);
+        ComponentTestHelper.SetInput(
+          comp, polyline.ToPolylineCurve(), 0);
+      } else {
+        var mesh = new Rhino.Geometry.Mesh();
```
Comment (DominikaLos) [inline GsaGHTests/3_Components/2_Geometry/CreateElement2dTests.cs:43]: why not old version? just new Mesh()?

Comment (DominikaLos) [reply GsaGHTests/3_Components/2_Geometry/CreateElement2dTests.cs:43]: the same for polyline if possible

**Review by DominikaLos — CHANGES_REQUESTED** [review]

File: GsaGHTests/3_Components/2_Geometry/CreateElement2dTests.cs (hunk 5 above)
Comment (SandeepArup) [reply GsaGHTests/3_Components/2_Geometry/CreateElement2dTests.cs]: It is same. Have removed namespace

File: GsaGH/Parameters/2_Geometry/GsaElement2dGoo.cs [hunk 6]
```diff
@@ -27,44 +27,62 @@ public class GsaElement2dGoo : GH_OasysGeometricGoo<GsaElement2d>, IGH_PreviewDa
...
       return false;
     }
 
     public override void DrawViewportMeshes(GH_PreviewMeshArgs args) {
-      if (Value != null && Value.Mesh != null) {
-        args.Pipeline.DrawMeshShaded(Value?.Mesh,
-        args.Material.Diffuse == Color.FromArgb(255, 150, 0, 0)
-          ? Colours.Element2dFace : Colours.Element2dFaceSelected);
+      if (Value == null) { return; }
+      double tolerance = Rhino.RhinoDoc.ActiveDoc.ModelAbsoluteTolerance;
```
[the remark this answers is not in the document]
Comment (SandeepArup) [reply GsaGH/Parameters/2_Geometry/GsaElement2dGoo.cs]: Removed

File: GsaGH/Components/0_Model/GetModelGeometry.cs [hunk 7]
```diff
@@ -617,16 +660,24 @@ protected override void SolveInternal(IGH_DataAccess data) {
 
         if (!((IGH_PreviewObject)Params.Output[5]).Hidden) {
           member2dKeys = results.Mem2ds.Select(item => item.Value.Id).ToList();
-
+          double tolerance = Rhino.RhinoDoc.ActiveDoc.ModelAbsoluteTolerance;
```
[the remark this answers is not in the document]
Comment (SandeepArup) [reply GsaGH/Components/0_Model/GetModelGeometry.cs]: removed

File: GsaGH/Parameters/2_Geometry/GsaElement2dGoo.cs [hunk 8]
```diff
@@ -27,44 +27,62 @@ public class GsaElement2dGoo : GH_OasysGeometricGoo<GsaElement2d>, IGH_PreviewDa
...
-      } else {
-        args.Pipeline.DrawMeshWires(Value.Mesh, Colours.Element2dEdgeSelected, 2);
-      }
-
       if (Value.Section3dPreview != null) {
-        if (args.Color == Color.FromArgb(255, 150, 0, 0)) {
-          args.Pipeline.DrawLines(Value.Section3dPreview.Outlines, Colours.Element2dEdge, 1);
-        } else {
-          args.Pipeline.DrawLines(Value.Section3dPreview.Outlines, Colours.Element2dEdgeSelected, 2);
+        args.Pipeline.DrawLines(Value.Section3dPreview.Outlines, args.Color == Color.FromArgb(255, 150, 0, 0) ? Colours.Element2dEdge : Colours.Element2dEdgeSelected, 1);
```
[the remark this answers is not in the document]
Comment (SandeepArup) [reply GsaGH/Parameters/2_Geometry/GsaElement2dGoo.cs]: assigned to variable

File: GsaGH/Parameters/2_Geometry/GsaElementFactory.cs (hunk 4 above)
Comment (SandeepArup) [reply GsaGH/Parameters/2_Geometry/GsaElementFactory.cs]: removed

File: GsaGH/Components/2_Geometry/Create2dElement.cs (hunk 3 above)
Comment (SandeepArup) [reply GsaGH/Components/2_Geometry/Create2dElement.cs]: @psarras Lets discuss this once you are back

File: GsaGHTests/1_BaseParameters/2_Geometry/GsaElement2dTest.cs [hunk 9]
```diff
@@ -386,5 +400,52 @@ private GsaElement2d CreateSampleElement2dWithQuad4Type() {
...
+      var pts = new Point3dList {
+       new Point3d(0, 0, 0),
+       new Point3d(2, 0, 0),
+       new Point3d(2, 2, 0),
+       new Point3d(0, 2, 0),
+     };
+
+      pts.Add(pts[0]);
+      var pol = new Polyline(pts);
+      var elem = new GsaElement2d(pol.ToPolylineCurve());
```
Comment (psarras) [inline GsaGHTests/1_BaseParameters/2_Geometry/GsaElement2dTest.cs:414]: This part is creating a GsaElement2D, we are doing this in multiple parts. Could we extract and re-use this? I noticed that sometimes you have different Point3ds, I don't understand which case are they testing?

if we need different points could we have the points as a specific list

```
var points = GetPlanarPoints();
var gsaElement2D = GetElement2dFromPoints(points)
```

File: GsaGHTests/1_BaseParameters/2_Geometry/GsaElement2dTest.cs [hunk 10]
```diff
@@ -386,5 +400,52 @@ private GsaElement2d CreateSampleElement2dWithQuad4Type() {
...
+        names.Add("Shahin");
+        elem.Prop2ds.Add(new GsaProperty2d(sectionId) {
+          ApiProp2d = new Prop2D(),
+        });
+      }
+
+      elem.ApiElements.SetMembers(groups);
+      elem.ApiElements.SetMembers(dummy);
+      elem.ApiElements.SetMembers(names);
+      return elem;
```
Comment (psarras) [inline GsaGHTests/1_BaseParameters/2_Geometry/GsaElement2dTest.cs:436]: The next Part, could we wrap it into a separate method? Are the "22" for group, dummy "true" significant for what we are testing or can we use the same? It seems like they are not.

I am noticing that sometimes we use "off" sometimes we don't could we add those as effects.

```
var element2d = GetElement2dFrom(mesh);
setDefaults(element2d);
applyOff(element2d);
```

File: GsaGHTests/1_BaseParameters/2_Geometry/GsaElement2dTest.cs [hunk 11]
```diff
@@ -386,5 +400,52 @@ private GsaElement2d CreateSampleElement2dWithQuad4Type() {
...
+    }
+
+    [Fact]
+    public void GetCenterPointsReturnsValidPointsForLoadPanel() {
+      GsaElement2d ele = CreateLoadPanel();
```
Comment (psarras) [inline GsaGHTests/1_BaseParameters/2_Geometry/GsaElement2dTest.cs:441]: where possible do not use appreviations

```
 GsaElement2d **elements** = CreateLoadPanel();
```

File: GsaGHTests/2_GooWrappers/GsaElement2dGooTests.cs [hunk 12]
```diff
@@ -1,32 +1,108 @@
...
   [Collection("GrasshopperFixture collection")]
   public class GsaElement2dGooTests {
-    [Fact]
-    public void GsaElement2dGooDrawViewportMeshesAndWiresTest() {
-      var comp = (Section3dPreviewComponent)CreateElement2dTests.ComponentMother();
+    [Theory]
+    [InlineData(true)]
+    [InlineData(false)]
+    public void GsaElement2dGooDrawViewportMeshesAndWiresTest(bool isLoadPanel) {
+      var comp = (Section3dPreviewComponent)CreateElement2dTests.ComponentMother(isLoadPanel, isLoadPanel);
```
Comment (psarras) [inline GsaGHTests/2_GooWrappers/GsaElement2dGooTests.cs:24]: Specify the type of the `ComponentMother` so we avoid casting.

Also move this common setup to the constructor

File: GsaGHTests/2_GooWrappers/GsaElement2dGooTests.cs [hunk 13]
```diff
@@ -1,32 +1,108 @@
...
+    public void GsaElement2dGooCastToCurveTest() {
+      var comp = (Section3dPreviewComponent)CreateElement2dTests.ComponentMotherLoadPanel();
+      var output = (GsaElement2dGoo)ComponentTestHelper.GetOutput(comp);
+      var curve = new GH_Curve();
+      Assert.True(output.CastTo(ref curve));
+      Assert.True(curve.IsValid);
+    }
+
+    [Fact]
+    public void MorphMeshTest() {
```
Comment (psarras) [inline GsaGHTests/2_GooWrappers/GsaElement2dGooTests.cs:61]: could we give a more descriptive name for these tests (Morph....) What are they meant to test?

for instance

`StretchingASection3dShouldOutputAValidMesh` 

Although I don't understand what we are testing, as the first part is always the same and the transformation we are changing (morph through stretch or translation) are Rhino functions, which we can assume they work.

**Review by psarras — CHANGES_REQUESTED** [review]

File: GsaGH/Helpers/Graphics/Colours.cs [hunk 14]
```diff
@@ -64,6 +66,27 @@ public static DisplayMaterial Element2dFaceSelected {
         return material;
       }
     }
+
+    public static DisplayMaterial Element2dFaceLP {
```
[the remark this answers is not in the document]
Comment (SandeepArup) [reply GsaGH/Helpers/Graphics/Colours.cs:70]: Transparency is diffrent

File: GsaGH/Helpers/Graphics/Colours.cs [hunk 15]
```diff
@@ -64,6 +66,27 @@ public static DisplayMaterial Element2dFaceSelected {
...
+      get {
+        var material = new DisplayMaterial {
```
Comment (DominikaLos) [inline GsaGH/Helpers/Graphics/Colours.cs:72]:
```suggestion
private static Color Element2dFace = Color.FromArgb(50, 195, 218, 219);

public static DisplayMaterial Element2dFaceLP {
      get {
        var material = new DisplayMaterial {
          Diffuse = Element2dFace,
          Transparency = 0.1,
        };
        return material;
      }
    }

    public static DisplayMaterial Element2dFaceSelectedLP {
      get {
        var material = new DisplayMaterial {
          Diffuse = Element2dFace,
          Transparency = 0.4,
        };
        return material;
      }
    }

```

File: GsaGH/Helpers/Graphics/Colours.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (DominikaLos) [reply GsaGH/Helpers/Graphics/Colours.cs:70]:
```suggestion
    private static Color Element2dFace = Color.FromArgb(50, 195, 218, 219);

    public static DisplayMaterial Element2dFaceLP {
      get {
        var material = new DisplayMaterial {
          Diffuse = Element2dFace,
          Transparency = 0.1,
        };
        return material;
      }
    }

    public static DisplayMaterial Element2dFaceSelectedLP {
      get {
        var material = new DisplayMaterial {
          Diffuse = Element2dFace,
          Transparency = 0.4,
        };
        return material;
      }
    }
```

File: GsaGH/Components/0_Model/GetModelGeometry.cs [hunk 16]
```diff
@@ -614,24 +575,22 @@ protected override void SolveInternal(IGH_DataAccess data) {
...
           member2dKeys = results.Mem2ds.Select(item => item.Value.Id).ToList();
-
-          var element2dsShaded = new ConcurrentBag<GsaElement2dGoo>();
-          var element2dsNotShaded = new ConcurrentBag<GsaElement2dGoo>();
           Parallel.ForEach(results.Elem2ds, elem => {
-            try {
-              int parent = elem.Value.ApiElements[0].ParentMember.Member;
-              if (parent > 0 && member2dKeys.Contains(parent)) {
-                element2dsShaded.Add(elem);
+            int parent = elem.Value.ApiElements[0].ParentMember.Member;
```
[the remark this answers is not in the document]
Comment (SandeepArup) [reply GsaGH/Components/0_Model/GetModelGeometry.cs:579]: Good suggestion...I just extended existing code. Will make appropriate changes.

File: GsaGH/Components/0_Model/GetModelGeometry.cs [hunk 17]
```diff
@@ -823,5 +782,47 @@ private void UpdateHiddenOutputs(bool nodeFilter, bool elementFilter, bool membe
...
+              }
+              break;
+            case Mesh mesh:
+              args.Display.DrawMeshShaded(mesh, Attributes.Selected ? Colours.Element2dFaceSelected : Colours.Element2dFace);
+              break;
+            default: break;
+          }
+        }
+      }
+    }
```
[the remark this answers is not in the document]
Comment (DominikaLos) [reply GsaGH/Components/0_Model/GetModelGeometry.cs]: ok

File: GsaGHTests/1_BaseParameters/2_Geometry/GsaElement2dTest.cs (hunk 9 above)
Comment (SandeepArup) [reply GsaGHTests/1_BaseParameters/2_Geometry/GsaElement2dTest.cs]: I have moved this in common place https://github.com/arup-group/GSA-Grasshopper/pull/695/files#diff-ff2742c0dcc2f3a0c413a11880749c0f04aca194050e993211639d8313fdf970

File: GsaGHTests/2_GooWrappers/GsaElement2dGooTests.cs (hunk 13 above)
Comment (SandeepArup) [reply GsaGHTests/2_GooWrappers/GsaElement2dGooTests.cs]: done

File: GsaGH/Components/0_Model/GetModelGeometry.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (SandeepArup) [reply GsaGH/Components/0_Model/GetModelGeometry.cs:579]: done

File: GsaGHTests/1_BaseParameters/2_Geometry/GsaElement2dTest.cs (hunk 11 above)
Comment (SandeepArup) [reply GsaGHTests/1_BaseParameters/2_Geometry/GsaElement2dTest.cs]: Tried to implement as much as possible related to this PR

**Review by psarras — APPROVED** [review]

**Pull request merged**